examples/render_pyramid_to_mbtiles.py

```diff
@@ -175,114 +116,115 @@ def add_tile(self, tile, data):
-    try:
-        render_tile(tile, sources, output)
-    except Exception:
-        logger.exception('Error while processing tile %s', tile)
-        raise
+    return catalog
 
 
-def render_tile(tile, sources, output):
-    for (type, transformation, format, ext) in RENDER_COMBINATIONS:
-
-        with Timer() as t:
-            (headers, data) = tiling.render_tile(
-                tile, sources, format=format, transformation=transformation)
+def render_tile_from_sources(tile, sources):
+    with Timer() as t:
+        (headers, data) = tiling.render_tile_from_sources(
+            tile,
+            sources,
+            format=PNG_FORMAT,
+            transformation=IMAGE_TRANSFORMATION)
 
-        logger.debug(
-            '(%02d/%06d/%06d) Took %0.3fs to render %s tile (%s bytes), Source: %s, Timers: %s',
-            tile.z, tile.x, tile.y, t.elapsed, type,
-            len(data),
-            headers.get('X-Imagery-Sources'),
-            headers.get('X-Timers'),
-        )
+    logger.debug(
+        '(%02d/%06d/%06d) Took %0.3fs to render tile (%s bytes), Source: %s, Timers: %s',
+        tile.z,
+        tile.x,
+        tile.y,
+        t.elapsed,
+        len(data),
+        headers.get('X-Imagery-Sources'),
+        headers.get('X-Timers'), )
 
-        MBTILES_POOL.apply_async(
-            write_to_mbtiles,
-            args=[type, tile, headers, data, output]
-        )
+    return headers, data
 
 
-def write_to_mbtiles(type, tile, headers, data, output):
+def write_to_mbtiles(tile, headers, data, output):
     try:
         with Timer() as t:
-            outputter = output.get(type)
-            outputter.add_tile(tile, data)
+            output.add_tile(tile, data)
 
         logger.debug(
-            '(%02d/%06d/%06d) Took %0.3fs to write %s tile to mbtiles://%s',
-            tile.z, tile.x, tile.y, t.elapsed, type,
-            outputter._filename,
-        )
-    except:
+            '(%02d/%06d/%06d) Took %0.3fs to write tile to mbtiles://%s',
+            tile.z,
+            tile.x,
+            tile.y,
+            t.elapsed,
+            output._filename, )
+    except Exception:
         logger.exception("Problem writing to mbtiles")
 
 
-def queue_tile(tile, max_zoom, sources, output):
-    queue_render(tile, sources, output)
+def queue_tile(executor, tile, max_zoom, sources, output):
+    queue_render(executor, tile, sources, output)
 
     if tile.z < max_zoom:
         for child in mercantile.children(tile):
-            queue_tile(child, max_zoom, sources, output)
+            queue_tile(executor, child, max_zoom, sources, output)
+
+
+def make_tile_done_callback(tile, output):
+    def tile_done_callback(tile_future):
+        if not tile_future.cancelled():
+            headers, data = tile_future.result()
 
+            logger.info(threading.current_thread())
+            # TODO this needs to occur on the main thread OR on a thread
+            # specifically for MBTiles
```

Comment:
@iandees thoughts on how to do this? Registering a done callback guarantees that it runs on a thread in the calling process, but not necessarily the same thread. I was thinking I could consume a generator that produces Futures, but I really want something that consumes **completed** Futures.

Putting MBTiles / SQLite on its own thread and queueing responses (from the done callback) could work; do you know of a higher-level way to create the worker than `threading`?

Reply:
I don't know of a higher level library to help with this stuff. I was going to suggest a queue. Something like they talk about in here: https://pymotw.com/2/multiprocessing/communication.html